<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Document-Source" data-toc-modified-id="Document-Source-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Document Source</a></span></li><li><span><a href="#Objectives" data-toc-modified-id="Objectives-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Objectives</a></span></li><li><span><a href="#Tools-&amp;-Concepts" data-toc-modified-id="Tools-&amp;-Concepts-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Tools &amp; Concepts</a></span><ul class="toc-item"><li><span><a href="#Jupyter" data-toc-modified-id="Jupyter-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Jupyter</a></span><ul class="toc-item"><li><span><a href="#@todo:-image,-link,-quoted-description" data-toc-modified-id="@todo:-image,-link,-quoted-description-3.1.1"><span class="toc-item-num">3.1.1&nbsp;&nbsp;</span>@todo: image, link, quoted description</a></span></li><li><span><a href="#Setup" data-toc-modified-id="Setup-3.1.2"><span class="toc-item-num">3.1.2&nbsp;&nbsp;</span>Setup</a></span><ul class="toc-item"><li><span><a href="#Displaying-results-in-the-notebook" data-toc-modified-id="Displaying-results-in-the-notebook-3.1.2.1"><span class="toc-item-num">3.1.2.1&nbsp;&nbsp;</span>Displaying results in the notebook</a></span></li><li><span><a href="#DIY-Helpers" data-toc-modified-id="DIY-Helpers-3.1.2.2"><span class="toc-item-num">3.1.2.2&nbsp;&nbsp;</span>DIY Helpers</a></span></li><li><span><a href="#Reading-Excel-files,-processing-data" data-toc-modified-id="Reading-Excel-files,-processing-data-3.1.2.3"><span class="toc-item-num">3.1.2.3&nbsp;&nbsp;</span>Reading Excel files, processing data</a></span></li></ul></li></ul></li><li><span><a href="#Python-Data-Structures" data-toc-modified-id="Python-Data-Structures-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Python Data Structures</a></span><ul class="toc-item"><li><span><a href="#Lists-&amp;-tuples" data-toc-modified-id="Lists-&amp;-tuples-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Lists &amp; tuples</a></span><ul class="toc-item"><li><span><a href="#Basic-definition-and-access" data-toc-modified-id="Basic-definition-and-access-3.2.1.1"><span class="toc-item-num">3.2.1.1&nbsp;&nbsp;</span>Basic definition and access</a></span></li><li><span><a href="#Slice-Sequences" data-toc-modified-id="Slice-Sequences-3.2.1.2"><span class="toc-item-num">3.2.1.2&nbsp;&nbsp;</span>Slice Sequences</a></span></li><li><span><a href="#Access-a-list-with-for" data-toc-modified-id="Access-a-list-with-for-3.2.1.3"><span class="toc-item-num">3.2.1.3&nbsp;&nbsp;</span>Access a <code>list</code> with <code>for</code></a></span></li><li><span><a href="#Define-a-list-with-for" data-toc-modified-id="Define-a-list-with-for-3.2.1.4"><span class="toc-item-num">3.2.1.4&nbsp;&nbsp;</span>Define a <code>list</code> with <code>for</code></a></span></li><li><span><a href="#Filter-a-list-with-if" data-toc-modified-id="Filter-a-list-with-if-3.2.1.5"><span class="toc-item-num">3.2.1.5&nbsp;&nbsp;</span>Filter a list with <code>if</code></a></span></li></ul></li><li><span><a href="#Sets" data-toc-modified-id="Sets-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Sets</a></span><ul class="toc-item"><li><span><a href="#Basic-definition-and-access" data-toc-modified-id="Basic-definition-and-access-3.2.2.1"><span class="toc-item-num">3.2.2.1&nbsp;&nbsp;</span>Basic definition and access</a></span></li><li><span><a href="#Set-Theory-Operations" data-toc-modified-id="Set-Theory-Operations-3.2.2.2"><span class="toc-item-num">3.2.2.2&nbsp;&nbsp;</span>Set Theory Operations</a></span></li></ul></li><li><span><a href="#Dictionaries" data-toc-modified-id="Dictionaries-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Dictionaries</a></span><ul class="toc-item"><li><span><a href="#Basic-definition-and-access" data-toc-modified-id="Basic-definition-and-access-3.2.3.1"><span class="toc-item-num">3.2.3.1&nbsp;&nbsp;</span>Basic definition and access</a></span></li><li><span><a href="#Using-for-to-access-a-dict" data-toc-modified-id="Using-for-to-access-a-dict-3.2.3.2"><span class="toc-item-num">3.2.3.2&nbsp;&nbsp;</span>Using <code>for</code> to access a <code>dict</code></a></span></li><li><span><a href="#Using-for-to-define-a-dict" data-toc-modified-id="Using-for-to-define-a-dict-3.2.3.3"><span class="toc-item-num">3.2.3.3&nbsp;&nbsp;</span>Using <code>for</code> to define a <code>dict</code></a></span></li><li><span><a href="#defaultdict:-an-Agreeable-Dictionary" data-toc-modified-id="defaultdict:-an-Agreeable-Dictionary-3.2.3.4"><span class="toc-item-num">3.2.3.4&nbsp;&nbsp;</span><code>defaultdict</code>: an Agreeable Dictionary</a></span></li></ul></li></ul></li><li><span><a href="#pandas" data-toc-modified-id="pandas-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span><em>pandas</em></a></span></li></ul></li><li><span><a href="#The-Excel-Workbook:-2017-CAM-data-from-iPads.xlxs" data-toc-modified-id="The-Excel-Workbook:-2017-CAM-data-from-iPads.xlxs-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>The Excel Workbook: <code>2017 CAM data from iPads.xlxs</code></a></span><ul class="toc-item"><li><span><a href="#Opening-an-Excel-File-with-pandas" data-toc-modified-id="Opening-an-Excel-File-with-pandas-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Opening an Excel File with <em>pandas</em></a></span></li><li><span><a href="#Worksheets-in-the-File,-as-DataFrames" data-toc-modified-id="Worksheets-in-the-File,-as-DataFrames-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Worksheets in the File, as DataFrames</a></span></li><li><span><a href="#Goal:-Unifying-the-DataFrames" data-toc-modified-id="Goal:-Unifying-the-DataFrames-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Goal: Unifying the DataFrames</a></span><ul class="toc-item"><li><span><a href="#pandas-functions-for-merging-DataFrames" data-toc-modified-id="pandas-functions-for-merging-DataFrames-4.3.1"><span class="toc-item-num">4.3.1&nbsp;&nbsp;</span><em>pandas</em> functions for merging DataFrames</a></span></li><li><span><a href="#Column-Name-Difference" data-toc-modified-id="Column-Name-Difference-4.3.2"><span class="toc-item-num">4.3.2&nbsp;&nbsp;</span>Column Name Difference</a></span><ul class="toc-item"><li><span><a href="#Visualize-the-Difference" data-toc-modified-id="Visualize-the-Difference-4.3.2.1"><span class="toc-item-num">4.3.2.1&nbsp;&nbsp;</span>Visualize the Difference</a></span></li><li><span><a href="#How-shall-we-solve-the-name-mismatch?" data-toc-modified-id="How-shall-we-solve-the-name-mismatch?-4.3.2.2"><span class="toc-item-num">4.3.2.2&nbsp;&nbsp;</span>How shall we solve the name mismatch?</a></span></li><li><span><a href="#Where-should-we-rename-the-columns?" data-toc-modified-id="Where-should-we-rename-the-columns?-4.3.2.3"><span class="toc-item-num">4.3.2.3&nbsp;&nbsp;</span>Where should we rename the columns?</a></span></li><li><span><a href="#Renaming-axis-labels-(columns-or-rows)-with-pandas" data-toc-modified-id="Renaming-axis-labels-(columns-or-rows)-with-pandas-4.3.2.4"><span class="toc-item-num">4.3.2.4&nbsp;&nbsp;</span>Renaming axis labels (columns or rows) with <em>pandas</em></a></span></li></ul></li><li><span><a href="#Finally-Uniting-Our-(Properly-Labeled)-DataFrames" data-toc-modified-id="Finally-Uniting-Our-(Properly-Labeled)-DataFrames-4.3.3"><span class="toc-item-num">4.3.3&nbsp;&nbsp;</span>Finally Uniting Our (Properly Labeled) DataFrames</a></span></li></ul></li><li><span><a href="#Goal:-Map-the-Column-Relationships" data-toc-modified-id="Goal:-Map-the-Column-Relationships-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Goal: Map the Column Relationships</a></span><ul class="toc-item"><li><span><a href="#Indexing-by-Repeated-Words" data-toc-modified-id="Indexing-by-Repeated-Words-4.4.1"><span class="toc-item-num">4.4.1&nbsp;&nbsp;</span>Indexing by Repeated Words</a></span></li><li><span><a href="#Split-the-Column-Names" data-toc-modified-id="Split-the-Column-Names-4.4.2"><span class="toc-item-num">4.4.2&nbsp;&nbsp;</span>Split the Column Names</a></span><ul class="toc-item"><li><span><a href="#The-Separator-String" data-toc-modified-id="The-Separator-String-4.4.2.1"><span class="toc-item-num">4.4.2.1&nbsp;&nbsp;</span>The Separator String</a></span></li><li><span><a href="#Resulting-Gain:-Expression-for-Splitting-Column-Names-into-Words" data-toc-modified-id="Resulting-Gain:-Expression-for-Splitting-Column-Names-into-Words-4.4.2.2"><span class="toc-item-num">4.4.2.2&nbsp;&nbsp;</span>Resulting Gain: Expression for Splitting Column Names into Words</a></span></li></ul></li><li><span><a href="#Build-a-Tree-of-Words" data-toc-modified-id="Build-a-Tree-of-Words-4.4.3"><span class="toc-item-num">4.4.3&nbsp;&nbsp;</span>Build a Tree of Words</a></span><ul class="toc-item"><li><span><a href="#Steps" data-toc-modified-id="Steps-4.4.3.1"><span class="toc-item-num">4.4.3.1&nbsp;&nbsp;</span>Steps</a></span></li><li><span><a href="#Define-Function:-split_column_to_tree" data-toc-modified-id="Define-Function:-split_column_to_tree-4.4.3.2"><span class="toc-item-num">4.4.3.2&nbsp;&nbsp;</span>Define Function: <code>split_column_to_tree</code></a></span></li><li><span><a href="#Use-that-Function-in-Another-Function" data-toc-modified-id="Use-that-Function-in-Another-Function-4.4.3.3"><span class="toc-item-num">4.4.3.3&nbsp;&nbsp;</span>Use that Function in Another Function</a></span></li><li><span><a href="#Visualize-the-Tree" data-toc-modified-id="Visualize-the-Tree-4.4.3.4"><span class="toc-item-num">4.4.3.4&nbsp;&nbsp;</span>Visualize the Tree</a></span></li><li><span><a href="#Refer-to-the-Data" data-toc-modified-id="Refer-to-the-Data-4.4.3.5"><span class="toc-item-num">4.4.3.5&nbsp;&nbsp;</span>Refer to the Data</a></span></li><li><span><a href="#Resulting-Gain" data-toc-modified-id="Resulting-Gain-4.4.3.6"><span class="toc-item-num">4.4.3.6&nbsp;&nbsp;</span>Resulting Gain</a></span></li></ul></li><li><span><a href="#Setup-Filters" data-toc-modified-id="Setup-Filters-4.4.4"><span class="toc-item-num">4.4.4&nbsp;&nbsp;</span>Setup Filters</a></span><ul class="toc-item"><li><span><a href="#Define-Functions" data-toc-modified-id="Define-Functions-4.4.4.1"><span class="toc-item-num">4.4.4.1&nbsp;&nbsp;</span>Define Functions</a></span></li><li><span><a href="#Improve-visualize_tree" data-toc-modified-id="Improve-visualize_tree-4.4.4.2"><span class="toc-item-num">4.4.4.2&nbsp;&nbsp;</span>Improve <code>visualize_tree</code></a></span></li></ul></li></ul></li><li><span><a href="#Putting-it-all-together" data-toc-modified-id="Putting-it-all-together-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Putting it all together</a></span><ul class="toc-item"><li><span><a href="#Accessing-data-from-any-section-&amp;-ignoring-irrelevant-rows" data-toc-modified-id="Accessing-data-from-any-section-&amp;-ignoring-irrelevant-rows-4.5.1"><span class="toc-item-num">4.5.1&nbsp;&nbsp;</span>Accessing data from any section &amp; ignoring irrelevant rows</a></span></li><li><span><a href="#Drop-Missing-Values" data-toc-modified-id="Drop-Missing-Values-4.5.2"><span class="toc-item-num">4.5.2&nbsp;&nbsp;</span>Drop Missing Values</a></span></li><li><span><a href="#Lookup-by-Condition" data-toc-modified-id="Lookup-by-Condition-4.5.3"><span class="toc-item-num">4.5.3&nbsp;&nbsp;</span>Lookup by Condition</a></span></li><li><span><a href="#Outputting-cleaner-data" data-toc-modified-id="Outputting-cleaner-data-4.5.4"><span class="toc-item-num">4.5.4&nbsp;&nbsp;</span>Outputting cleaner data</a></span></li><li><span><a href="#Analyzing-sections-and-subsections" data-toc-modified-id="Analyzing-sections-and-subsections-4.5.5"><span class="toc-item-num">4.5.5&nbsp;&nbsp;</span>Analyzing sections and subsections</a></span></li></ul></li></ul></li><li><span><a href="#Source-App-Data-Structure" data-toc-modified-id="Source-App-Data-Structure-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Source App Data Structure</a></span><ul class="toc-item"><li><span><a href="#Section:-clients" data-toc-modified-id="Section:-clients-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Section: <code>clients</code></a></span></li><li><span><a href="#Section:-observers" data-toc-modified-id="Section:-observers-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Section: <code>observers</code></a></span></li><li><span><a href="#Section:-fields" data-toc-modified-id="Section:-fields-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>Section: <code>fields</code></a></span><ul class="toc-item"><li><span><a href="#fields-Subsection:-client" data-toc-modified-id="fields-Subsection:-client-5.3.1"><span class="toc-item-num">5.3.1&nbsp;&nbsp;</span><code>fields</code> Subsection: <code>client</code></a></span></li><li><span><a href="#fields-Subsection:-oSets" data-toc-modified-id="fields-Subsection:-oSets-5.3.2"><span class="toc-item-num">5.3.2&nbsp;&nbsp;</span><code>fields</code> Subsection: <code>oSets</code></a></span></li><li><span><a href="#oSets-Subsection:-oPoints" data-toc-modified-id="oSets-Subsection:-oPoints-5.3.3"><span class="toc-item-num">5.3.3&nbsp;&nbsp;</span><code>oSets</code> Subsection: <code>oPoints</code></a></span><ul class="toc-item"><li><span><a href="#oPoints-Subsection:-location" data-toc-modified-id="oPoints-Subsection:-location-5.3.3.1"><span class="toc-item-num">5.3.3.1&nbsp;&nbsp;</span><code>oPoints</code> Subsection: <code>location</code></a></span></li><li><span><a href="#oPoints-Subsection:-observations" data-toc-modified-id="oPoints-Subsection:-observations-5.3.3.2"><span class="toc-item-num">5.3.3.2&nbsp;&nbsp;</span><code>oPoints</code> Subsection: <code>observations</code></a></span></li></ul></li></ul></li></ul></li></ul></div>

# Document Source

* You may view or download this notebook on [NBViewer], though it won't be interactive.
* You may interact with this notebook on [mybinder.org], though changes won't be saved for very long.
* The [source] for this notebook is online at GitHub.com, where it can be viewed or downloaded. Instructions on how to use this notebook on your own computer are in the [README].

[![Binder](https://mybinder.org/badge.svg)][mybinder.org]

[README]: https://github.com/devvyn/aafc-field-data/blob/master/README.md
[source]: https://github.com/devvyn/aafc-field-data/blob/master/notebook/Using%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb
[mybinder.org]: https://mybinder.org/v2/gh/devvyn/aafc-field-data/master?filepath=notebook%2FUsing%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb
[NBViewer]: https://nbviewer.jupyter.org/github/devvyn/aafc-field-data/blob/master/notebook/Using%20Python%20%26%20pandas%20to%20Analyze%20an%20Excel%20File.ipynb

# Objectives

* Get to know our analytical tools:
  - Jupyter Notebook - interactive document editor
  - Python - expressive programming language
  - _pandas_ - data science framework

* Merge two Excel worksheets into one DataFrame

* Query the data

* Export the query results to Excel

* Document all steps, to ensure these results are reproducible

# Tools & Concepts

## Jupyter

### @todo: image, link, quoted description

- This document was written in a web browser connected to a Jupyter Notebook server.
- All the code samples shown were actually run on the server and the output saved in the document.

### Setup

To work with the source data, we need to be able to:

* open an Excel file
* display results of intermediate processing in the notebook
* save results to an Excel file

Most functions to accomplish our goals have been published already. We only need to import them.

#### Displaying results in the notebook

You can use the regular [`print()`] function in Python, but you'll always get plain text output. [`display()`] can do plain text as well as HTML, images, audio, etc.

I found I didn't need to import the [`display()`] method on new versions of Jupyter Notebook, but since the display function for isn't always imported by default in some implementations, and I also want to use `IPython.core.display.HTML()`:

[`print()`]: https://docs.python.org/3/library/functions.html#print 
[`display()`]: http://ipython.readthedocs.io/en/stable/api/generated/IPython.display.html

In [25]:
from IPython.core.display import *

#### DIY Helpers

Here are simple helpers to make headings and other embellishments easy. There are probably libraries already to make HTML tags, but this should suffice, without importing a large library. In general, these methods are unnecessary, but I'll be using them later on in this document to make code output more understandable.

In [26]:
def html_wrap(content, element="span", attributes_str=None):
    """ Convenience method for wrapping any string in any HTML tag. """
    if attributes_str:
        element_open = f"{element} {attributes_str}"
    else:
        element_open = element
    return f"<{element_open}>{content}</{element}>"


def make_heading(content, level=5):
    """ Convenience method for wrapping any string in an HTML header tag. """
    element = f"h{level}"
    return html_wrap(content, element)


def heading(*args, **kwargs):
    """ Call IPython.core.display.HTML on the results of the make_heading
    function, so users don't have to repeatedly do `HTML(heading(…))` """
    return HTML(make_heading(*args, **kwargs))

#### Reading Excel files, processing data

[_pandas_] is really good with columnar data, such as the kind of spreadsheets Excel works with. In fact, [_pandas_] has the ability to read and write Excel files, as long as the [`xlrd`] package is installed in Python.

The Jupyter server I used for writing this notebook already has [_pandas_] and [`xlrd`], but if you need to install them, you can run this command in a Jupyter Notebook cell, like so:

[_pandas_]: https://pandas.pydata.org
[`xlrd`]: https://xlrd.readthedocs.io/en/latest/

In [27]:
# Install xlrd to the Python environment, if it's not already installed.
!pip install xlrd pandas

Depending on how your Jupyter host is deployed, you may be able to add [`xlrd`] and [`pandas`][_pandas_] to [`requirements.txt`] or [`environment.yml`], as appropriate for your configuration. I [configured] my Jupyter server for use with [repo2docker]. In particular, I used [`requirements.txt`] and [`postBuild`], because even though [Conda] is usually better than [`pip`] at enabling Jupyter extensions, the installer would hang a lot, wasting my time, whenever I used [`environment.yml`].

[_pandas_]: https://pandas.pydata.org
[`xlrd`]: https://xlrd.readthedocs.io/en/latest/
[repo2docker]: http://repo2docker.readthedocs.io/en/latest/index.html
[configured]: http://repo2docker.readthedocs.io/en/latest/usage.html#locating-and-composing-configuration-files
[`requirements.txt`]: http://repo2docker.readthedocs.io/en/latest/usage.html#requirements-txt
[`postBuild`]: http://repo2docker.readthedocs.io/en/latest/usage.html#postbuild
[`environment.yml`]: http://repo2docker.readthedocs.io/en/latest/usage.html#environment-yml
[Conda]: https://conda.io/docs/
[`pip`]: https://pip.pypa.io/en/stable/user_guide/

In [28]:
import pandas

If you prefer, you can install this from the regular command line shell on your Jupyter host by running the `pip` command without the exclamation mark.

## Python Data Structures

- list
- tuple
- slicing
- expressions:
  - for … in
  - if
- set
- dict
- defaultdict

### Lists & tuples

A [`list`] in Python is simply a sequence of objects, which can be numbers, text, or anything else. It uses square brackets.

```python
> my_list = [1, 20, 4004]
> my_list[2] = 1001
> my_list[2]
1001
```

[`list`]: https://docs.python.org/3/library/stdtypes.html#list

A [`tuple`] is almost identical to a `list`, except it can't be edited after it's defined. You define it using round parentheses, instead of square brackets.

```python
> my_tuple = (1, 20, 4004)
> my_tuple[2] = 1001
TypeError: 'tuple' object does not support item assignment
> my_tuple[2]
4004
```

[`tuple`]: https://docs.python.org/3/library/stdtypes.html#tuples

#### Basic definition and access

##### Define a tuple or list:

In [29]:
my_list = [1, 2, 3, "four"]

a_tiny_list = [8]

In [30]:
my_tuple = (1, 2, 3, "four", ("nested 1", "nested 2", a_tiny_list), 30, 0)

##### Nested tuples & lists

In [31]:
another_list = [my_list, 5, 6, 7, my_tuple]

In [32]:
another_list

[[1, 2, 3, 'four'],
 5,
 6,
 7,
 (1, 2, 3, 'four', ('nested 1', 'nested 2', [8]), 30, 0)]

##### Access items in a `list` or `tuple`:

There's an "item zero" because 0 is the origin, just like in Cartesian coordinates.

In [33]:
my_list[0]

1

In [34]:
my_tuple[0]

1

The index sequence continues as you might expect:

In [35]:
my_list[1]

2

In [36]:
my_list[2]

3

In [37]:
my_list[3]

'four'

If you attempt to access a non existent item by using an index that's out of range, you'll receive an error.

In [38]:
my_list[4]

IndexError: list index out of range

You can, however, use negative numbers! Negative numbers are assumed to be relative to the upper boundary of the range. -1 refers to the _last_ item, -2 to the _second last_, and so on.

In [39]:
my_list[-1]

'four'

In [40]:
my_tuple[-2]

30

##### Adding items to a list:

In [41]:
a_tiny_list

[8]

In [42]:
a_tiny_list.append("another item")

In [43]:
a_tiny_list

[8, 'another item']

#### Slice Sequences

If you want to choose a portion of a list, you can use the colon character to separate the start and end index numbers you want to use.

In [44]:
my_list[0:2]  # 0, 1 but not 2

[1, 2]

In [45]:
my_tuple[5:7]  # 5, 6 but not 7

(30, 0)

You can omit the first or last number in your slice to use the natural upper or lower boundary:

In [46]:
my_list[:2]  # until just before 2

[1, 2]

In [47]:
my_tuple[5:]  # 5 and after

(30, 0)

#### Access a `list` with `for`

In [48]:
for item in my_list:
    print(item)

1
2
3
four


In [49]:
for item in another_list:
    print(item)

[1, 2, 3, 'four']
5
6
7
(1, 2, 3, 'four', ('nested 1', 'nested 2', [8, 'another item']), 30, 0)


####  Define a `list` with `for`

Evaluate an expression based on every item in a list:

In [50]:
my_new_list = [f"I have {item}!"
               for item in my_list]
my_new_list

['I have 1!', 'I have 2!', 'I have 3!', 'I have four!']

#### Filter a list with `if`

In [51]:
my_filtered_list = [item for item in my_list 
                    if item == "four" or item > 2]
my_filtered_list

[3, 'four']

### Sets

A [`set`] in Python in similar to a list, except there is no order of the items. Like a list, a set can be changed after it's defined.

[`set`]: https://docs.python.org/3/library/stdtypes.html#set

#### Basic definition and access

In [52]:
my_set = {12, 24, 48, 96}

In [53]:
my_set

{12, 24, 48, 96}

In [54]:
my_set.add(96 * 2)

In [55]:
my_set

{12, 24, 48, 96, 192}

##### Make a set from an iterable, such as a list or tuple

In [56]:
my_other_list = [0, 25, 48]

In [57]:
another_set = set(my_other_list)
another_set

{0, 25, 48}

#### Set Theory Operations

Because Python is explicitly treating [`set`] objects as the kind of sets discussed in [set theory], you can do operations on them. 

[set theory]: https://en.wikipedia.org/wiki/Set_theory
[`set`]: https://docs.python.org/3/library/stdtypes.html#set

##### Difference / Exclusion

In [58]:
my_set = {12, 24, 48, 96, 192}
another_set = {0, 25, 48}

In [59]:
my_set - another_set

{12, 24, 96, 192}

In [60]:
another_set - my_set

{0, 25}

##### Union & Intersection

In [61]:
my_set = {12, 24, 48, 96, 192}
another_set = {0, 25, 48}

In [62]:
my_set | another_set

{0, 12, 24, 25, 48, 96, 192}

In [63]:
my_set & another_set

{48}

##### Symmetric Difference

In [64]:
my_set = {12, 24, 48, 96, 192}
another_set = {0, 25, 48}

In [65]:
my_set ^ another_set

{0, 12, 24, 25, 96, 192}

In [66]:
set.symmetric_difference(my_set, another_set)

{0, 12, 24, 25, 96, 192}

### Dictionaries

Dictionaries in Python are just a collection of named things. The things can be another dictionary, a string, a number, or whatver. Even the names of the things don't necessarily have to be words—they can be numbers, for example.

#### Basic definition and access

##### Define a dictionary:

In [67]:
my_dictionary = {
    'one': 1,
    2: 'two',
    "green": "I like colour green.",
    "another dictionary": {"more stuff": 1024,
                           'even more stuff': 2048},
    "a list": my_list
}

##### Access stored values:

In [68]:
my_dictionary["one"]

1

In [69]:
my_dictionary[2]

'two'

In [70]:
my_dictionary["another dictionary"]["even more stuff"]

2048

Note that any attempt to use a non-existant key results in a Python exception called a `KeyError`. This is helpful if you make a human error, or unexpectedly find keys missing. (There _are_ ways around it.)

In [71]:
my_dictionary["some key that doesn't exist"]

KeyError: "some key that doesn't exist"

##### Assign values to existing keys or new keys:

In [72]:
my_dictionary["two"] = my_dictionary[2]
my_dictionary[2] = "fourteen"

In [73]:
my_dictionary["some key that didn't exist before"] = "Some value that didn't exist before"

In [74]:
my_dictionary.keys()

dict_keys(['one', 2, 'green', 'another dictionary', 'a list', 'two', "some key that didn't exist before"])

#### Using `for` to access a `dict`

##### Dictionary iterables

Python's `for` statements are handy for doing something to each item in a collection. Dictionaries are a type of collection, but since they have keys and values, you need to specify which you want. To address this, there are three handy methods that all dictionaries have:

- [`keys()`] - Only the key names
- [`values()`] - Only the values
- [`items()`] - Pairs of keys and values

[`keys()`]: https://docs.python.org/3/library/stdtypes.html#dict.keys
[`values()`]: https://docs.python.org/3/library/stdtypes.html#dict.values
[`items()`]: https://docs.python.org/3/library/stdtypes.html#dict.items



##### `keys()`

In [75]:
for key in my_dictionary.keys():
    display(key)

'one'

2

'green'

'another dictionary'

'a list'

'two'

"some key that didn't exist before"

##### `values()`

In [76]:
for value in my_dictionary.values():
    display(value)

1

'fourteen'

'I like colour green.'

{'even more stuff': 2048, 'more stuff': 1024}

[1, 2, 3, 'four']

'two'

"Some value that didn't exist before"

##### `items()`

In [77]:
for item in my_dictionary.items():
    display(item)

('one', 1)

(2, 'fourteen')

('green', 'I like colour green.')

('another dictionary', {'even more stuff': 2048, 'more stuff': 1024})

('a list', [1, 2, 3, 'four'])

('two', 'two')

("some key that didn't exist before", "Some value that didn't exist before")

The "item" returned by `items()` is always a tuple of two values. You can name them anything you want, but here's the gist of the idea:

`(key, value)`

In [78]:
my_dictionary = {
    2: 'two',
    "another dictionary": {"more stuff": 1024,
                           'even more stuff': 2048},
    "a list": my_list
}
for key, value in my_dictionary.items():
    print(f"Key: {key}\n"
          f"Value: {value}\n")

Key: 2
Value: two

Key: another dictionary
Value: {'more stuff': 1024, 'even more stuff': 2048}

Key: a list
Value: [1, 2, 3, 'four']



#### Using `for` to define a `dict`

With a list of items, it's often useful to make a dictionary with keys or values based those items.

For example, let's say we have this filtered list expression:

In [79]:
[item + 100
 for item in my_list if item != "four"]


[101, 102, 103]

The syntax for dictionary creation with `for` is almost the same as that of lists, but like in standard dictionary definitions, you'd use `key: value` for the assignment. You can use expressions to transform the keys as well as the values, if you want.


In [80]:
{f"Key #{item}": item + 100
 for item in my_list if item != "four"}

{'Key #1': 101, 'Key #2': 102, 'Key #3': 103}

This is the same as using the regular assignment syntax in a regular `for` loop:

In [81]:
# Equivalent to:
# {f"Key #{item}": item + 100
#  for item in my_list if item != "four"}
my_new_dict = {}
for item in my_list:
    if item != "four":
        my_new_dict[f"Key #{item}"] = item + 100

You should use whichever format is clearest for your situation.

#### `defaultdict`: an Agreeable Dictionary

[`defaultdict`] is like a regular dictionary, except it doesn't complain if you try to access a key that doesn't exist yet—it just adds it.

[`defaultdict`]: https://docs.python.org/3/library/collections.html#collections.defaultdict

Remember how a regular `dict` handles invalid keys? Suppose we attempt to append to a list that we think is stored in a certain key, even though the dictionary is empty:


In [82]:
my_dictionary = dict()

In [83]:
my_dictionary["some key that doesn't exist"].append("an item")
my_dictionary["some key that doesn't exist"].append("another item")

KeyError: "some key that doesn't exist"

Can't append to a list object that doesn't exist!

If it were a [`defaultdict`], set to make `list` objects by default:

[`defaultdict`]: https://docs.python.org/3/library/collections.html#collections.defaultdict

In [84]:
from collections import defaultdict

In [85]:
my_defaultdict = defaultdict(list)

In [86]:
my_defaultdict["some key that doesn't exist"].append("an item")
my_defaultdict["some key that doesn't exist"].append("another item")

In [87]:
my_defaultdict

defaultdict(list, {"some key that doesn't exist": ['an item', 'another item']})

## _pandas_

@todo: intro to _pandas_

* Series
* Index
* DataFrame
* I/O

# The Excel Workbook: `2017 CAM data from iPads.xlxs`

In [88]:
# The file I'm interested in parsing for cleanup:
file_path = "./data/real/2017/CAM data from iPads/2017 CAM data from iPads.xlsx"

## Opening an Excel File with _pandas_

In [89]:
data_file = pandas.ExcelFile(file_path)

If you get an error here, it may be due to a missing module called `xlrd`. The [setup instructions] will tell you how to install it.

[setup instructions]: #Reading-Excel-files,-processing-data

## Worksheets in the File, as DataFrames

In [90]:
data_file_sheet_names = sorted(data_file.sheet_names)

In [91]:
data_file_sheet_names

['2017 CAM data Erl',
 '2017 CAM iPad data Tyler',
 'Combined iPad 2017 CAM data',
 'schema (WIP reverse engineer)']

I'm only interested in first two, which should have all the data. The other sheets are either unfinished or unrelated to the task at hand.

In [92]:
cam_sheet_names = data_file_sheet_names[:2]

In [93]:
cam_sheet_names

['2017 CAM data Erl', '2017 CAM iPad data Tyler']

In [94]:
def last_word(string, word_separator=' '):
    words = string.split(word_separator)
    return words[-1]

In [95]:
last_word("hey, there are words here")

'here'

Let's make a dictionary of DataFrames from all the sheets, using the last word as the name. If you're familiar with [lists] in Python, you may remember that an index of -1 means the last item, as in the largest possible index number. We're using `for` to define the `dict` based on items in `cam_sheet_names`. _pandas_ DataFrames can be made by calling the `parse` function from the _pandas_ file object, given a sheet name.

[lists]: https://docs.python.org/3/tutorial/datastructures.html#lists

In [96]:
sheets = {last_word(sheet_name): data_file.parse(sheet_name)
          for sheet_name in cam_sheet_names}

In [97]:
sheets.keys()

dict_keys(['Erl', 'Tyler'])

## Goal: Unifying the DataFrames

Now that we have a convenient list of sheets that are loaded as _pandas_ DataFrames, we can work toward merging them into one. Once they're merged, we can process the data more easily from a single DataFrame.

### _pandas_ functions for merging DataFrames

_pandas_ has multiple methods for combining datasets:

- [`concat`]: The concatenate method has an option to ignore row numbers, which has the effect of gluing each dataframe to the bottom of the previous one.
- [`append`]: Append would be almost equivalent, but always updates the first DataFrame or Series it's given.
- [`merge`]: Merge is more for database-style relational merges.

[`concat`]: https://pandas.pydata.org/pandas-docs/stable/merging.html
[`append`]: https://pandas.pydata.org/pandas-docs/stable/merging.html#concatenating-using-append
[`merge`]: https://pandas.pydata.org/pandas-docs/stable/merging.html#database-style-dataframe-joining-merging

I'm thinking about [`append`], but I like to have the option of creating a new DataFrame rather than overwriting anything, because the results are easier to repeat.

From reading up on all three methods, it looks to me like `concat` will suffice, here, as long as the column names are identical.

Let's see how close that concatenation method gets us:

In [98]:
pandas.concat(sheets).info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)
Data columns (total 56 columns):
clients__company                                                    1 non-null object
clients__displayText                                                1 non-null object
clients__fname                                                      1 non-null object
clients__lname                                                      1 non-null object
clients__name                                                       1 non-null object
fields__client__company                                             21 non-null object
fields__client__displayText                                         21 non-null object
fields__client__fname                                               21 non-null object
fields__client__lname                                               21 non-null object
fields__client__name                                                21 non-null object
fields__crop

We can see the MultiIndex size in the info: 

>MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)

Which means 4690 "rows", if this were a spreadsheet. 

Compare to the total row count for all sheets, by using Python's function `len`:

In [99]:
data_frame_lengths = [len(sheet) for sheet in sheets.values()]
display(data_frame_lengths)
sum(data_frame_lengths)

[910, 3780]

4690

>Data columns (total 56 columns)

The `info()` readout for the concatenation dry run tells us about columns.

In [100]:
[len(sheet.columns) for sheet in sheets.values()]

[50, 50]

The column counts for each source worksheet.

Pretty good, so far, except for some extra columns due to variations in the column names—I see `growthStage` and `growthStage Zadoks`, as well as some aphid names amongst the `a1__number`, `a2__number`, `a3__number` group.

* Unchanging number of rows
  + Merge result: Good 😃
    + no redundancy in rows
    + no missing rows
* Rising number of columns
  - Merge result: Sub-optimal 😡
    - redundancy in columns
    - incomplete merge

### Column Name Difference

#### Visualize the Difference

By looking at the column list from the dry run, it's already fairly obvious what the naming problem is. But just to demonstrate how to use Python to calculate differences, let's look again at our DataFrames column names.

For our frames:

In [101]:
for key, sheet in sheets.items():
    print(f"{key}, {type(sheet)!r}")

Erl, <class 'pandas.core.frame.DataFrame'>
Tyler, <class 'pandas.core.frame.DataFrame'>


In [102]:
print('\n'.join((f"{key}, {type(sheet)!r}" for key, sheet in sheets.items())))

Erl, <class 'pandas.core.frame.DataFrame'>
Tyler, <class 'pandas.core.frame.DataFrame'>


To list the names of columns from *all* sheets that don't have a match (symmetric difference), we can use [`set.symmetric_difference`], which is also written as the `^` symbol demonstrated in the [introduction to sets].

[introduction to sets]: #Sets
[`set.symmetric_difference`]: https://docs.python.org/3/library/stdtypes.html#frozenset.symmetric_difference

In [103]:
column_list = sheets['Erl'].columns ^ sheets['Tyler'].columns
display(column_list)

Index(['fields__oSets__growthStage', 'fields__oSets__growthStage Zadoks',
       'fields__oSets__oPoints__observations__a1__number',
       'fields__oSets__oPoints__observations__a1__number EGA',
       'fields__oSets__oPoints__observations__a2__number',
       'fields__oSets__oPoints__observations__a2__number BCO',
       'fields__oSets__oPoints__observations__a3__number',
       'fields__oSets__oPoints__observations__a3__number Greenbug',
       'fields__oSets__oPoints__observations__anum',
       'fields__oSets__oPoints__observations__anum TotalAPhids',
       'fields__oSets__oPoints__observations__eVnum',
       'fields__oSets__oPoints__observations__eVnum Natural enemy totals'],
      dtype='object')

Here are the problematic parts:

`fields__oSets__growthStage
fields__oSets__growthStage `**`Zadoks`**`<
fields__oSets__oPoints__observations__a1__number
fields__oSets__oPoints__observations__a1__number `**`EGA`**`<
fields__oSets__oPoints__observations__a2__number
fields__oSets__oPoints__observations__a2__number `**`BCO`**`<
fields__oSets__oPoints__observations__a3__number
fields__oSets__oPoints__observations__a3__number `**`Greenbug`**`<
fields__oSets__oPoints__observations__anum
fields__oSets__oPoints__observations__anum `**`TotalAPhids`**`<
fields__oSets__oPoints__observations__eVnum
fields__oSets__oPoints__observations__eVnum `**`Natural enemy totals`**`<`

Before we concatenate, let's standardize on the names from the more regular DataFrame. But which sheet is the one to fix? Which sheet has `fields__oSets__growthStage Zadoks` (two "words") instead of `fields__oSets__growthStage` (one "word")?

Let's make sure of what we're about to do, by checking for any column names have multiple "words":

In [104]:
sheets_with_bad_column_names = {sheet_name: sheet for sheet_name, sheet in sheets.items() 
                                if any(len(column.split()) > 1 for column in sheet.columns)}

In [105]:
display(heading('Bad:'),
        set(sheets_with_bad_column_names.keys()))

{'Erl'}

There's now a `sheets_with_bad_column_names` list we can use when we fix that.

##### To Confirm:

All the sheets, minus the bad ones:

In [106]:
display(heading('Good:'),
        sheets.keys() - sheets_with_bad_column_names.keys())

{'Tyler'}

Okay, "Tyler" has the column names we prefer, "Erl" does not. Duly noted.

#### How shall we solve the name mismatch?

Our options:

- rename the columns (somewhere)
- ignore the column names of the bad sheet, merge based on order of columns

If we try to do a special merge, ignoring the column names, we have to worry about the _precise order_ of those columns instead of relying on the names. Since we'd have to spend time checking—and possibly rearranging—the column sequence, we might as well fix the names. We can do so while documenting the steps, for posterity.

#### Where should we rename the columns?

Possibilities:

- fix the spreadsheet document in Excel, re-import with _pandas_ and move ahead as if this had never happened
- correct the names in memory only, on the fly, with _pandas_

Since this notebook you're reading has already demonstrated the problem, and we're using _pandas_ anyway, let's go ahead and solve it here.

#### Renaming axis labels (columns or rows) with _pandas_

_pandas_ has a [`rename`] method which lets us:

- Apply a transform function to the DataFrame columns that are named after our worksheet columns. 
- Use a dictionary to explicitly map each column to its respective new name.

Either way, we'll need a function or expression for changing each name. Since the bad names merely have extra words tacked onto the end, let's just split the name and use the first "word".

[`rename`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.rename.html

In Python, we get the first item of a sequence by using index zero. (As you may remember from the earlier [introduction to lists].)

[introduction to lists]: #Lists-&-tuples

In [107]:
def first_word(string, word_separator=' '):
    """ Split string into words (by space character), return first word. """
    words = string.split(word_separator)
    return words[0]

Since the function was so easy to make, let's not bother with a dictionary. For the sake of visualizing the changes, let's report on changes in a dry run before using the _pandas_ rename function. You certainly don't have to do anything like this, but I enjoy working in Python.

In [108]:
report = []  # For the lines of the report which we'll display later.
for sheet_name, frame in sheets_with_bad_column_names.items():
    report.append(make_heading(sheet_name, level=5))  # heading for each frame
    report_lines = []  # the body of the report for each DataFrame
    for before in frame.columns:
        after = first_word(before)
        if after != before:  # not equal
            content = f"{before} <br> &rarr; {after}"  # old --> new
            attributes = "style='font-weight: bold'"  # CSS
        else:
            content = before
            attributes = None
        report_lines.append(html_wrap(content, 'li', attributes))  # HTML list item
    report.append(html_wrap(''.join(report_lines), 'ol'))  # HTML ordered list
    
# join all the report lines together and display as HTML
display(HTML(''.join(report)))

Looks perfect! Actually change the names:

In [109]:
for sheet in sheets_with_bad_column_names.values():
    sheet.rename(mapper=first_word, axis='columns', inplace=True)

##### Developer notes:

As mentioned before, we know there's only one sheet, so the `for` loop isn't absolutely necessary, but it's still a good habit when dealing with reusable scripts that can work on batches of datasets. (Sometimes other people find your work useful, and sometimes you'll revisit your work to copy something that succeeded in the past.)

### Finally Uniting Our (Properly Labeled) DataFrames

In [110]:
data = pandas.concat(sheets)

In [111]:
data.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 4690 entries, (Erl, 0) to (Tyler, 3779)
Columns: 50 entries, clients__company to observers
dtypes: datetime64[ns](1), float64(27), object(22)
memory usage: 1.8+ MB


Splendid! That's what we expected. 😁

## Goal: Map the Column Relationships

### Indexing by Repeated Words

In [112]:
data.columns

Index(['clients__company', 'clients__displayText', 'clients__fname',
       'clients__lname', 'clients__name', 'fields__client__company',
       'fields__client__displayText', 'fields__client__fname',
       'fields__client__lname', 'fields__client__name', 'fields__crop',
       'fields__date', 'fields__desc', 'fields__image', 'fields__name',
       'fields__oSets__completeSets', 'fields__oSets__date',
       'fields__oSets__dateCompare', 'fields__oSets__desc',
       'fields__oSets__growthStage', 'fields__oSets__oPoints__id',
       'fields__oSets__oPoints__location__coords__accuracy',
       'fields__oSets__oPoints__location__coords__altitude',
       'fields__oSets__oPoints__location__coords__altitudeAccuracy',
       'fields__oSets__oPoints__location__coords__heading',
       'fields__oSets__oPoints__location__coords__latitude',
       'fields__oSets__oPoints__location__coords__longitude',
       'fields__oSets__oPoints__location__coords__speed',
       'fields__oSets__oPoints__loc

We're dealing with a lot of repetition in column names, which indicates relationships to each other which we can use to group our columns. The more often a word is repeated, the more it outranks its fellow words — or the higher its seniority in the family tree, so to speak.

Let's use a nested dictionary in Python to map out a relational hierarchy.

Also, _pandas_ has a concept of [indexing hierarchically], which may help group columns together — we can probably use the dictionary to help us add more indices to the DataFrame later, if we need.

[indexing hierarchically]: https://pandas.pydata.org/pandas-docs/stable/advanced.html

### Split the Column Names

Our column names follow a pattern: words separated by two underscores.

If we turn a column name into a list of words, we can start grouping those words based on the word that came before them.

For example, these column names:

```
fields__name
fields__oSets__oPoints__observations__name
fields__oSets__oPoints__observations__anum
fields__oSets__oPoints__observations__enum
```

Would become a tree of words:

```
        _ name                                  _ name
       /                                       /
fields --> oSets --> oPoints --> observations --- anum
                                               \_ enum
```

#### The Separator String

Here's what we're dealing with.

In [113]:
display(sheets['Erl'].columns[3])

'fields__client__lname'

Between words is `__` — save that string of underscore characters for later reference.

In [114]:
SEPARATOR = '__'

##### Python's `split()` method for strings:

In [115]:
column_name = sheets['Erl'].columns[3]
column_name.split(SEPARATOR)

['fields', 'client', 'lname']

We can boldly use that expression whenever we please, without defining any functions.

In [116]:
f"So many fields about {column_name.split(SEPARATOR)[0]}, I tell you!"

'So many fields about fields, I tell you!'

#### Resulting Gain: Expression for Splitting Column Names into Words

In [117]:
for column_name in sheets['Erl'].columns:
    display(column_name.split(SEPARATOR))

['fields', 'client', 'company']

['fields', 'client', 'name']

['fields', 'client', 'fname']

['fields', 'client', 'lname']

['fields', 'client', 'displayText']

['fields', 'name']

['fields', 'crop']

['fields', 'desc']

['fields', 'image']

['fields', 'date']

['fields', 'oSets', 'date']

['fields', 'oSets', 'dateCompare']

['fields', 'oSets', 'growthStage']

['fields', 'oSets', 'desc']

['fields', 'oSets', 'obsName']

['fields', 'oSets', 'totalSets']

['fields', 'oSets', 'completeSets']

['fields', 'oSets', 'results']

['fields', 'oSets', 'oPoints', 'id']

['fields', 'oSets', 'oPoints', 'name']

['fields', 'oSets', 'oPoints', 'observations', 'id']

['fields', 'oSets', 'oPoints', 'observations', 'name']

['fields', 'oSets', 'oPoints', 'observations', 'enum']

['fields', 'oSets', 'oPoints', 'observations', 'eVnum']

['fields', 'oSets', 'oPoints', 'observations', 'anum']

['fields', 'oSets', 'oPoints', 'observations', 'a1', 'number']

['fields', 'oSets', 'oPoints', 'observations', 'a2', 'number']

['fields', 'oSets', 'oPoints', 'observations', 'a3', 'number']

['fields', 'oSets', 'oPoints', 'observations', 'disabled']

['fields', 'oSets', 'oPoints', 'observations', 'complete']

['fields', 'oSets', 'oPoints', 'observations', '|']

['fields', 'oSets', 'oPoints', 'observations', '|', 'number']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'latitude']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'longitude']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'accuracy']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'altitude']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'heading']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'speed']

['fields', 'oSets', 'oPoints', 'location', 'coords', 'altitudeAccuracy']

['fields', 'oSets', 'oPoints', 'location', 'timestamp']

['fields', 'oSets', 'totalA1']

['fields', 'oSets', 'totalA2']

['fields', 'oSets', 'totalA3']

['fields', 'oSets', 'totalA4']

['clients', 'company']

['clients', 'name']

['clients', 'fname']

['clients', 'lname']

['clients', 'displayText']

['observers']

Looks good to me. Some words aren't separated, like `dateCompare` and `growthStage`, but those shouldn't pose a problem because they're nicely grouped with related columns.

### Build a Tree of Words

#### Steps

Here's the strategy for building a tree from those lists of words for each column name, with data:

0. Given:
   - a _pandas_ `DataFrame`
   - a column name and separator string
   - a `defaultdict` that makes more `defaultdict` objects if you just ask
     for the key
0. Set a pointer to the new `defaultdict`, which will be root of the tree, for starters.
0. Split the column name apart, using the given separator string.
0. Now, walk through the words in order from left to right:
   0. access whichever `defaultdict` the pointer is currently referring to
   0. lookup a key based the current word
   0. update the pointer to the `defaultdict` for that key (whether it existed before or not)
0. On the last word, update the pointer:
   + store a reference to the data for that `DataFrame` column.

#### Define Function: `split_column_to_tree`

In [118]:
def split_column_to_tree(data_frame, column_name, tree=None,
                         separator=SEPARATOR, data_key=SEPARATOR):
    """Split the column names like "fields__oSets__oPoints__id" into groupings
    of keys so that related keys are easy to find, such as 
    columns['fields']['oSets']['oPoints']['id']. This produces a tree of
    column name segments, with references to actual data at the ends.
    """
    if tree is None:
        def nested_dict():
            return defaultdict(nested_dict)
        tree = nested_dict()
    pointer = tree
    for word in str(column_name).split(separator):
        pointer = pointer[word]
    pointer[data_key] = data_frame[column_name]
    return tree


##### The Key to Real Data:

Because there's always the chance that a node might have data as well as more nodes under it, we'll store the reference to data with a key having a name that can't possibly be a word: the separator!

#### Use that Function in Another Function

By passing the partially constructed tree back into our function for re-use, we can accumulate all the columns together.

In [119]:
def make_column_tree(data_frame, separator=SEPARATOR, data_key=SEPARATOR):
    tree = None  # initialize the variable, otherwise we can't reference it
    for name in data.columns:
        tree = split_column_to_tree(data_frame, name, tree, separator, data_key)
    return tree

In [120]:
columns = make_column_tree(data)

##### Check Column Groups

If that worked as planned, there should be a list of the first segments of all the column names in that sheet:

In [121]:
columns.keys()

dict_keys(['clients', 'fields', 'observers'])

Continuing deeper, more segments that share a common prefix:

In [122]:
fields_node = columns['fields']
fields_node.keys()

dict_keys(['client', 'crop', 'date', 'desc', 'image', 'name', 'oSets'])

In [123]:
sets_node = fields_node['oSets']
sets_node.keys()

dict_keys(['completeSets', 'date', 'dateCompare', 'desc', 'growthStage', 'oPoints', 'obsName', 'results', 'totalA1', 'totalA2', 'totalA3', 'totalA4', 'totalSets'])

#### Visualize the Tree

Browsing the column structure by using the dictionary accessors is a bit tedious. It's great for machines, but my human brain has a bit of trouble holding it all in my head when I see it only one level at a time.

We can see the structure of the hierarchy in a dictionary by using a function like this:

In [124]:
def visualize_tree(tree, depth=1, maximum_depth=None, data_key=SEPARATOR):
    if isinstance(tree, (dict, defaultdict)):
        for key, val in tree.items():
            if key != data_key:
                print(".   " * (depth - 1) + key)
                if maximum_depth is None or depth <= maximum_depth:
                    visualize_tree(val, depth + 1, maximum_depth, data_key)

It may seem strange to see math symbols used with text, but Python understands what we want: to repeat the text if we multiply it, to append if we add.

Here's the output for the whole tree:

In [125]:
visualize_tree(columns)

clients
.   company
.   displayText
.   fname
.   lname
.   name
fields
.   client
.   .   company
.   .   displayText
.   .   fname
.   .   lname
.   .   name
.   crop
.   date
.   desc
.   image
.   name
.   oSets
.   .   completeSets
.   .   date
.   .   dateCompare
.   .   desc
.   .   growthStage
.   .   oPoints
.   .   .   id
.   .   .   location
.   .   .   .   coords
.   .   .   .   .   accuracy
.   .   .   .   .   altitude
.   .   .   .   .   altitudeAccuracy
.   .   .   .   .   heading
.   .   .   .   .   latitude
.   .   .   .   .   longitude
.   .   .   .   .   speed
.   .   .   .   timestamp
.   .   .   name
.   .   .   observations
.   .   .   .   a1
.   .   .   .   .   number
.   .   .   .   a2
.   .   .   .   .   number
.   .   .   .   a3
.   .   .   .   .   number
.   .   .   .   anum
.   .   .   .   complete
.   .   .   .   disabled
.   .   .   .   eVnum
.   .   .   .   enum
.   .   .   .   id
.   .   .   .   name
.   .   .   .   |
.   .   .   .   .   number
.   .   o

Using this new function, we can also limit the depth, even to zero, which is the equivalent of `column_dictionary.keys()`:

In [126]:
columns.keys()

dict_keys(['clients', 'fields', 'observers'])

In [127]:
visualize_tree(columns, maximum_depth=0)

clients
fields
observers


Or, just a shallow peek into the next level:

In [128]:
visualize_tree(columns, maximum_depth=1)

clients
.   company
.   displayText
.   fname
.   lname
.   name
fields
.   client
.   crop
.   date
.   desc
.   image
.   name
.   oSets
observers


#### Refer to the Data

Earlier, we declared a `SEPARATOR` string, then decided to also use that as the key for storing a reference to data at the end of each branch in our tree index. For example, the `date` of the set:

In [129]:
sets_node = columns['fields']['oSets']
sets_node['date'].keys()

dict_keys(['__'])

Since we're not using that separator string anymore, let's give our data key a better name.

In [130]:
DATA = SEPARATOR

And we can check for that any time by using Python's `in` operator:

In [131]:
DATA in sets_node['date']

True

We can look at some data now!

In [132]:
sets_node['date'][DATA].head()

Erl  0    2017-08-02T13:12:09.542
     1                        NaN
     2                        NaN
     3                        NaN
     4                        NaN
Name: fields__oSets__date, dtype: object

There are many missing values, represented by `NaN`. There are ways to skip over them, but let's come back to that problem later when we're dealing with entire sections of the DataFrame.

#### Resulting Gain

Now, every column can be found in two places:

- Built-in _pandas_ DataFrame methods and properties
  - `data['fields__name'].describe()`
  - `data['fields__client__name'].unique()`
- Our index tree
  - `columns['fields']['name']['__'].describe()`
  - `columns['fields']['client']['name']['__'].unique()`

In concrete terms, the data came from the real world, and each column of data pertains to a particular thing, place, or time. The column names reflect those particulars. Now, during analysis in Jupyter Notebook, we can use Python's well-known dictionary access methods to directly query relationships between columns.

Soon, we'll take advantage of this structure to make expressions which intuitively select the relevent columns for any given section.

### Setup Filters

#### Define Functions

As soon as we want to read from groups of columns based on which batch of rows we're reading from, we'll want to easily produce a list of applicable columns for any given row. 

Unfortunately, we can't simply ignore the columns with missing values, because the app also uses the absence of a value to indicate a value of zero. This is logically a practice that loses information, and this may be an example of why it's generally unwise. We'll have to work around the developers' choice.

Instead, we can get lists of columns from any node in our index tree by using methods like `keys()` and `items()`. From there, we can follow the keys down to the next node, and so on, all the way to the data reference. All we have to do is figure out which keys point to data and which point to more nodes, and we can use `for` in Python to make a list of whatever we want.

For detecting data that we've placed in the hierarchy of columns, we can look for the special key that we chose earlier, which we ultimately named `DATA`.

`isinstance` is a built-in Python function which checks if an object is a dictionary, list, or whatever type you specify and returns `True` or `False`.

In [133]:
# DATA = SEPARATOR = '__'
def has_data(node, data_key=DATA):
    """ We know node has data if it has a key that's the data key. """
    return isinstance(node, dict) and data_key in node.keys()

We can easily use `for` and `has_data` together to build a filtered dictionary from any node.

Here's a function which looks at all items in a node and filters out any items that aren't data containers:

In [134]:
def child_data(node):
    """ Filter children that have data. """
    return {parent_key: child for parent_key, child in node.items()
            if isinstance(node, dict) and has_data(child)}

It's easy to assume that any item in a node that isn't a data container is therefore a node leading to more nodes. However, that isn't a valid assumption, because we've already seen that there are columns that disprove that case:

- `fields__oSets__oPoints__observations__|`
- `fields__oSets__oPoints__observations__|__number`

We'll have to check whether the item also has children that aren't _merely_ a data container.

It's a bit trickier to also check if a data container is also an intermediate node, but it can be done.

Here's the logic:

- If a node has more than one key:
  - Because there can only be one data key, any other keys must something else:
    - Assume it is a child node because we aren't expecting anything else.
- If a node has one key:
  - Check if that key is the data key:
    - If yes:
      - That can't possibly be a child node, because it's a data reference.
    - If no:
      - Assume it is a child node because we aren't expecting anything else.

In [135]:
def child_children(node):
    """ Filter children that have children. We know an child has children if
    it has any values that are dictionaries with at least one key that isn't
    a data reference. """
    return {parent_key: child for parent_key, child in node.items()
            if isinstance(child, dict) and len(child) > 1 or len(child) == 1 and not has_data(child)}

In [136]:
def descendants(node):
    """ Recursively return all children."""
    # @todo
    pass

In [137]:
points_node = sets_node['oPoints']
observations_node = points_node['observations']

In [138]:
display(heading('children with children:'),
        child_children(observations_node).keys())

dict_keys(['a1', 'a2', 'a3', '|'])

In [139]:
display(heading('children with data:'),
        child_data(observations_node).keys())

dict_keys(['anum', 'complete', 'disabled', 'eVnum', 'enum', 'id', 'name', '|'])

In [140]:
# With Python sets, the ampersand (`&`) is the set intersetion operator.
display(heading('children with children & data:'),
        set(child_children(observations_node).keys()) & set(child_data(observations_node).keys()))

{'|'}

Oh? Children and data under `|`? What could it be?

In [141]:
display(observations_node['|'][DATA].dropna().head(18))

Erl  5     e1
     6     e2
     7     e3
     8     e4
     9     e5
     10    e6
     11    e7
     12    e8
     13    e9
     19    e1
     20    e2
     21    e3
     22    e4
     23    e5
     24    e6
     25    e7
     26    e8
     27    e9
Name: fields__oSets__oPoints__observations__|, dtype: object

In [142]:
display(observations_node['|']['number'][DATA].dropna().head(18))

Erl  90     5.0
     94     1.0
     96     2.0
     104    5.0
     110    1.0
     118    4.0
     132    2.0
     136    1.0
     145    1.0
     152    1.0
     160    1.0
     166    1.0
     502    2.0
     509    1.0
     544    1.0
     558    2.0
     572    7.0
     586    6.0
Name: fields__oSets__oPoints__observations__|__number, dtype: float64

Based on my experience with using the app, I think it has something to do with natural enemy counts, because that's the only other information gathered by the app that hasn't been accounted for in the well-named columns.

As far as I've been able to discern, the source app's data doesn't distinguish between a count of zero and unreported (missing) data points. That's why the `observations__|__number` data (which doesn't necessarily exist) isn't always present on the same rows as all the `observations__|` labels.

Later on, we'll need to dig up the actual list of naturaly enemies and match them to the corresponding "e" labels, `e1` through `e9`.

#### Improve `visualize_tree`

Now that we have filters for helping us traverse the tree, let's take a moment to improve `visualize_tree`:

In [143]:
def visualize_tree(tree, depth=1, maximum_depth=None, data_key=DATA,
                   show_leaf_nodes=True):
    if isinstance(tree, (dict, defaultdict)):
        if show_leaf_nodes:
            nodes = tree
        else:
            nodes = child_children(tree)
        for key, val in nodes.items():
            if key != data_key:
                if has_data(val):
                    prefix = '* '
                else:
                    prefix = ''
                if child_children(val) or child_data(val):
                    suffix = '/'
                else:
                    suffix = ''
                print(".   " * (depth - 1) + f"{prefix}{key}{suffix}")
                if maximum_depth is None or depth <= maximum_depth:
                    visualize_tree(val, depth + 1, maximum_depth, data_key,
                                   show_leaf_nodes)

Now we can visualize the structure in terms of what's at the ends of the branches (leaf nodes) or all descendants. The additional punctuation also makes it more obvious which nodes have data values. For example, the node at `fields__oSets__oPoints__observations__|`, which is both a data node and a branch.

In [144]:
visualize_tree(observations_node, show_leaf_nodes=False)

a1/
a2/
a3/
* |/


In [145]:
visualize_tree(observations_node, show_leaf_nodes=True)

a1/
.   * number
a2/
.   * number
a3/
.   * number
* anum
* complete
* disabled
* eVnum
* enum
* id
* name
* |/
.   * number


## Putting it all together

### Accessing data from any section & ignoring irrelevant rows

Let's take a look at the columns containing data about observation sets.

We'll use `has_data` to filter our `sets_node` so we can express a list of names of data-containing columns about `fields__oSets`.

In [146]:
sets_columns_names = [column[DATA].name for column in child_data(sets_node).values()]
display(sets_columns_names)

['fields__oSets__completeSets',
 'fields__oSets__date',
 'fields__oSets__dateCompare',
 'fields__oSets__desc',
 'fields__oSets__growthStage',
 'fields__oSets__obsName',
 'fields__oSets__results',
 'fields__oSets__totalA1',
 'fields__oSets__totalA2',
 'fields__oSets__totalA3',
 'fields__oSets__totalA4',
 'fields__oSets__totalSets']

Passing that list to _pandas_, we should get exactly which columns of data we need:

In [147]:
display(data[sets_columns_names].head())

fields__oSets__completeSets      fields__oSets__date  \
Erl 0                          0.0  2017-08-02T13:12:09.542   
    1                          NaN                      NaN   
    2                          NaN                      NaN   
    3                          NaN                      NaN   
    4                          NaN                      NaN   

      fields__oSets__dateCompare  fields__oSets__desc  \
Erl 0                 2017-08-02                  NaN   
    1                        NaT                  NaN   
    2                        NaT                  NaN   
    3                        NaT                  NaN   
    4                        NaT                  NaN   

       fields__oSets__growthStage fields__oSets__obsName  \
Erl 0                         7.0                  Tyler   
    1                         NaN                    NaN   
    2                         NaN                    NaN   
    3                         NaN                    NaN   
    4                         NaN                    NaN   

      fields__oSets__results  fields__oSets__totalA1  fields__oSets__totalA2  \
Erl 0                    NaN                     NaN                     NaN   
    1                    NaN                     NaN                     NaN   
    2                    NaN                     NaN                     NaN   
    3                    NaN                     NaN                     NaN   
    4                    NaN                     NaN                     NaN   

       fields__oSets__totalA3  fields__oSets__totalA4  \
Erl 0                     NaN                     NaN   
    1                     NaN                     NaN   
    2                     NaN                     NaN   
    3                     NaN                     NaN   
    4                     NaN                     NaN   

       fields__oSets__totalSets  
Erl 0                       1.0  
    1                       NaN  
    2                       NaN  
    3                       NaN  
    4                       NaN

### Drop Missing Values

Many, many empty cells. If we want to skip over blanks as we scan down a column, _pandas_ has a [`dropna()`] method for that. Let's use `head` to peek at the first bit of that:

[`dropna()`]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.dropna.html

In [148]:
sets_node['date'][DATA].dropna().head(11)

Erl    0      2017-08-02T13:12:09.542
       70     2017-08-09T09:25:11.710
       140    2017-08-09T10:06:25.480
       210    2017-08-09T11:21:01.555
       350    2017-08-09T11:37:20.862
       490    2017-08-22T15:42:05.751
       560    2017-08-17T11:12:02.820
       700    2017-08-17T13:06:30.183
       840    2017-08-22T16:02:50.682
Tyler  0      2017-07-14T12:31:24.194
       70     2017-07-18T10:31:22.263
Name: fields__oSets__date, dtype: object

So much better!

We now have enough information to find the beginning of chunks in the sheet by scanning for non-blank cells: we have those row numbers for each of the non-null records of that column, where related columns will also reveal data we need for that section.

### Lookup by Condition

Once again, the problem we faced at the outset. How about skipping irrelevant rows? We fixed this earlier with `dropna()`, but this time we're operating on some columns that _may_ be null, plus certain ones that _must not_ be null.

Let's use the date column as the crucial record upon which we'll predicate our filter, because it should never be null. 

_pandas_ has a way to filter on conditions, which is sometimes called [boolean indexing] because the condition is either `True` or `False`. In this case, _pandas_ expects the actual column data object itself (Series), rather than just the name of the column to check. We already have that in the hierarchy from earlier:

[boolean indexing]: https://pandas.pydata.org/pandas-docs/stable/10min.html#boolean-indexing

In [149]:
date_column = sets_node['date'][DATA]
type(date_column)

pandas.core.series.Series

Passing a check for whether the value in the Series [is not null] to _pandas_, as well as specifying the column names we want to get:

[is not null]: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.notna.html

In [150]:
sets = data[date_column.notnull()][sets_columns_names]
display(sets.head(15))

fields__oSets__completeSets      fields__oSets__date  \
Erl   0                            0.0  2017-08-02T13:12:09.542   
      70                           1.0  2017-08-09T09:25:11.710   
      140                          1.0  2017-08-09T10:06:25.480   
      210                          2.0  2017-08-09T11:21:01.555   
      350                          2.0  2017-08-09T11:37:20.862   
      490                          1.0  2017-08-22T15:42:05.751   
      560                          2.0  2017-08-17T11:12:02.820   
      700                          2.0  2017-08-17T13:06:30.183   
      840                          1.0  2017-08-22T16:02:50.682   
Tyler 0                            0.0  2017-07-14T12:31:24.194   
      70                           1.0  2017-07-18T10:31:22.263   
      140                          1.0  2017-07-28T13:05:44.673   
      210                          1.0  2017-08-01T15:27:40.174   
      280                          1.0  2017-08-04T10:18:29.654   
      350                          1.0  2017-07-18T10:51:30.195   

          fields__oSets__dateCompare  fields__oSets__desc  \
Erl   0                   2017-08-02                  NaN   
      70                  2017-08-09                  NaN   
      140                 2017-08-09                  NaN   
      210                 2017-08-09                  NaN   
      350                 2017-08-09                  NaN   
      490                 2017-08-22                  NaN   
      560                 2017-08-17                  NaN   
      700                 2017-08-17                  NaN   
      840                 2017-08-22                  NaN   
Tyler 0                   2017-07-14                  NaN   
      70                  2017-07-18                  NaN   
      140                 2017-07-28                  NaN   
      210                 2017-08-01                  NaN   
      280                 2017-08-04                  NaN   
      350                 2017-07-18                  NaN   

           fields__oSets__growthStage fields__oSets__obsName  \
Erl   0                           7.0                  Tyler   
      70                          8.0                  Tyler   
      140                         7.0                  Tyler   
      210                         9.0                  Stean   
      350                         8.0                  Stean   
      490                         8.0                  Mikki   
      560                         8.0              Gabrielle   
      700                         9.0                  Stean   
      840                         8.0                  Mikki   
Tyler 0                           6.0                  Tyler   
      70                          6.0                  Tyler   
      140                         8.0                  Mikki   
      210                         8.0                  Stean   
      280                         8.5                  Stean   
      350                         6.0                  Tyler   

          fields__oSets__results  fields__oSets__totalA1  \
Erl   0                      NaN                     NaN   
      70               RESULTS.5                   164.0   
      140              RESULTS.5                    66.0   
      210              RESULTS.1                     0.0   
      350              RESULTS.1                     5.0   
      490              RESULTS.5                     1.0   
      560              RESULTS.1                   169.0   
      700              RESULTS.1                    78.0   
      840              RESULTS.5                   187.0   
Tyler 0                      NaN                     NaN   
      70               RESULTS.5                     8.0   
      140              RESULTS.5                    37.0   
      210              RESULTS.5                    52.0   
      280              RESULTS.5                    47.0   
      350              RESULTS.5    

Now, we can repeat this technique to get all the other data sections!

### Outputting cleaner data

To finish up, we should refer to that visualization of grouped columns, and start building a repeatable set of steps to grab the data from each section and relate them to each other before continuing on to the next set of nested sections in the DataFrame. As we grab sections and their related sub-sections, we can toss them into another DataFrame that will ultimately represent the same information in a cleaner form.

Because we're doing this programmatically, we can apply those repeatable steps as functions, in a loop. No manual slogging required. Hit play and relax, it'll be done before you can say "ahhh yeee"!

### Analyzing sections and subsections

In [151]:
visualize_tree(columns)

clients/
.   * company
.   * displayText
.   * fname
.   * lname
.   * name
fields/
.   client/
.   .   * company
.   .   * displayText
.   .   * fname
.   .   * lname
.   .   * name
.   * crop
.   * date
.   * desc
.   * image
.   * name
.   oSets/
.   .   * completeSets
.   .   * date
.   .   * dateCompare
.   .   * desc
.   .   * growthStage
.   .   oPoints/
.   .   .   * id
.   .   .   location/
.   .   .   .   coords/
.   .   .   .   .   * accuracy
.   .   .   .   .   * altitude
.   .   .   .   .   * altitudeAccuracy
.   .   .   .   .   * heading
.   .   .   .   .   * latitude
.   .   .   .   .   * longitude
.   .   .   .   .   * speed
.   .   .   .   * timestamp
.   .   .   * name
.   .   .   observations/
.   .   .   .   a1/
.   .   .   .   .   * number
.   .   .   .   a2/
.   .   .   .   .   * number
.   .   .   .   a3/
.   .   .   .   .   * number
.   .   .   .   * anum
.   .   .   .   * complete
.   .   .   .   * disabled
.   .   .   .   * eVnum
.   .   .   .   * enum
.   .  

In [152]:
pandas.DataFrame(data={"Legend Symbol": ('*', '/'),
                       "Meaning": ("Data", "More Keys")})

Legend Symbol    Meaning
0             *       Data
1             /  More Keys

For each section, I'd like to open with a summary. In my mind, this would include a picture of what's at the node. If any of these exist, let's see:

- Description of this data series
- Names of all related data series grouped here
- Names of all subsections that go deeper

Rather than repeating the same commands for each section, let's make a function that does all that, so we can just call it once for each section.

In [153]:
def summarize_node(node):
    reports = {"Data nodes": child_data(node),
               "Subsections": child_children(node)}
    report_frames = [
        pandas.Series(data=list(value.keys()), name=key)
        for key, value in reports.items()
        if value  # (not empty)
    ]
    if has_data(node):
        display(node[DATA].describe().to_frame())  # looks nicer as frame
    for series in report_frames:
        display(series.to_frame().T)  # easier to read in landscape orientation


# Source App Data Structure

## Section: `clients`

In [154]:
clients_node = columns['clients']
summarize_node(clients_node)

0            1      2      3     4
Data nodes  company  displayText  fname  lname  name

## Section: `observers`

In [155]:
observers_node = columns['observers']
summarize_node(observers_node)

observers
count          3
unique         3
top        Stean
freq           1

## Section: `fields`

In [156]:
fields_node = columns['fields']
visualize_tree(fields_node, show_leaf_nodes=False)
summarize_node(fields_node)

client/
oSets/
.   oPoints/
.   .   location/
.   .   .   coords/
.   .   observations/
.   .   .   a1/
.   .   .   a2/
.   .   .   a3/
.   .   .   * |/


0     1     2      3     4
Data nodes  crop  date  desc  image  name

0      1
Subsections  client  oSets

### `fields` Subsection: `client`

In [157]:
field_client_node = fields_node['client']
summarize_node(field_client_node)

0            1      2      3     4
Data nodes  company  displayText  fname  lname  name

### `fields` Subsection: `oSets`

In [158]:
osets_node = fields_node['oSets']
visualize_tree(osets_node, show_leaf_nodes=False)
summarize_node(osets_node)

oPoints/
.   location/
.   .   coords/
.   observations/
.   .   a1/
.   .   a2/
.   .   a3/
.   .   * |/


0     1            2     3            4        5   \
Data nodes  completeSets  date  dateCompare  desc  growthStage  obsName   

                 6        7        8        9        10         11  
Data nodes  results  totalA1  totalA2  totalA3  totalA4  totalSets

0
Subsections  oPoints

### `oSets` Subsection: `oPoints`

In [159]:
opoints_node = osets_node['oPoints']
visualize_tree(opoints_node, show_leaf_nodes=False)
summarize_node(opoints_node)

location/
.   coords/
observations/
.   a1/
.   a2/
.   a3/
.   * |/


0     1
Data nodes  id  name

0             1
Subsections  location  observations

#### `oPoints` Subsection: `location`

In [160]:
location_node = opoints_node['location']
visualize_tree(location_node)
summarize_node(location_node)

coords/
.   * accuracy
.   * altitude
.   * altitudeAccuracy
.   * heading
.   * latitude
.   * longitude
.   * speed
* timestamp


0
Data nodes  timestamp

0
Subsections  coords

##### `location` Subsection: `coords`

In [161]:
coords_node = location_node['coords']
summarize_node(coords_node)

0         1                 2        3         4  \
Data nodes  accuracy  altitude  altitudeAccuracy  heading  latitude   

                    5      6  
Data nodes  longitude  speed

#### `oPoints` Subsection: `observations`

In [162]:
observations_node = opoints_node['observations']
visualize_tree(observations_node)
summarize_node(observations_node)

a1/
.   * number
a2/
.   * number
a3/
.   * number
* anum
* complete
* disabled
* eVnum
* enum
* id
* name
* |/
.   * number


0         1         2      3     4   5     6  7
Data nodes  anum  complete  disabled  eVnum  enum  id  name  |

0   1   2  3
Subsections  a1  a2  a3  |

##### `observations` Subsections: `a1`, `a2`, `a3`

In [163]:
summarize_node(observations_node['a1'])

0
Data nodes  number

##### `observations` Subsection: `|` (natural enemies)

In [164]:
natural_enemies_node = observations_node['|']
summarize_node(natural_enemies_node)

fields__oSets__oPoints__observations__|
count                                     3015
unique                                       9
top                                         e9
freq                                       335

0
Data nodes  number

As far as I've been able to discern, the source app's data doesn't distinguish between a count of zero and unreported (missing) data points. That's why the `observations__|__number` data (which doesn't necessarily exist) isn't always present on the same rows as all the `observations__|` labels.

Later on, we'll need to dig up the actual list of naturaly enemies and match them to the corresponding "e" labels, `e1` through `e9`.

All possible values of `|`, which are the identities of the natural enemies:

In [165]:
pandas.DataFrame(observations_node['|'][DATA].dropna().unique()).T

0   1   2   3   4   5   6   7   8
0  e1  e2  e3  e4  e5  e6  e7  e8  e9